### ISSUE TITLE
Recaptcha v3 not working. Other recaptcha plugins work with the same keys

### ISSUE THREAD

**Issue by MarsWarrior** [Automattic/WP-Job-Manager#2850]
**Describe the bug**
A clear and concise description of what the bug is. Please be as descriptive as possible; issues lacking detail, or for any other reason than to report a bug, may be closed without action.

**To Reproduce**
I did create a v3 key and enabled v3 on the Job Manager configuration screen. I also enabled recaptcha in the Applications screen.

I can see that code is added in the inspector:

**Screenshots**
![image](https://github.com/user-attachments/assets/3c108576-2520-4a61-954a-fbe4487f69d1)

However:
- There is NO badge visible (other plugins like WP Captcha show a badge and function properly) Note: This seems logical as the backend PHP code does the verification if I look into the software.
- There is NO activity in the Google v3 console
- Bots are able to submit a form about every second!

**Expected behavior**
- Recaptcha works

**Isolating the problem (mark completed items with an [x]):**
- [partially. can't disable all plugins without database issues] I have deactivated other plugins and confirmed this bug occurs when only WP Job Manager plugin is active.
- [x] This bug happens with a default WordPress theme active.
- [x] I can reproduce this bug consistently using the steps above. The captcha never works!

**WordPress Environment**
- WordPress Version : v6.6.1
- WP Job Manager Version: v2.4.0 / WP Job Manager Applications v3.2.0
- PHP Version : 8.x
- Other important details:

**Comment by erricgunawan** [comment]
Perhaps related:

https://wordpress.org/support/topic/recaptchav3-2/#post-18112728

**Comment by samrueby** [comment]
I believe I found the cause of this issue. https://wordpress.org/support/topic/bug-in-displaying-recpatcha/

**Comment by igotdes** [comment]
Reported in #9442448-zen

**Comment by Gnodesign** [comment]
The reCaptcha does not appear on the initial page load. However, if the user submits the form without filling a required field (the category field), the page reloads with an error, and the reCaptcha loads correctly. This issue occurs because the reCaptcha script loads too late, as @samrueby correctly pointed out.

**Comment by samrueby** [comment]
Hi @Gnodesign ,

I'm trying to gauge how quickly this bug might be resolved. It's affecting our clients. 

I see there's no assignees. Does that indicate there's currently no plan to resolve this and we should be looking for alternative solutions?

Thank you.

**Comment by MarsWarrior** [comment]
> Hi [@Gnodesign](https://github.com/Gnodesign) ,
> 
> I'm trying to gauge how quickly this bug might be resolved. It's affecting our clients.
> 
> I see there's no assignees. Does that indicate there's currently no plan to resolve this and we should be looking for alternative solutions?
> 
> Thank you.

Same here. Clients buy WP Job Manager but don't see any fixes. Form submission is mostly disabled due to this bug, as any bot can submit a form with a non-working Captcha.

**Comment by Gnodesign** [comment]
Hi @samrueby,

The development has slowed down, and their team is currently focused on other projects, from what they told me. They still have plans for the plugin but no clear timeline for when they'll resume the development. I submitted several PRs a few months ago, and none have been reviewed. Addressing this issue ourselves seems like the best approach for now.

I have hCaptcha support on the roadmap for my Cariera theme, which is built on WPJM, but there's no ETA yet. If this issue remains unresolved by then, I may look into fixing it as well.

**Comment by samrueby** [comment]
Thank you for your response @Gnodesign so that we can plan accordingly!

**Comment by Gnodesign** [comment]
The issue occurs because the `wp_enqueue_scripts` hook has already fired by the time the `WP_Job_Manager_Recaptcha->enqueue_scripts()` is trying to run. The reCAPTCHA script loads correctly on default themes but fails to enqueue on other themes. You can fix the issue by changing this line:
https://github.com/Automattic/WP-Job-Manager/blob/41d5048c9cf6c4d2636c878419added8979fcdc6/includes/class-wp-job-manager-recaptcha.php#L80

to this:
```
// Check if wp_enqueue_scripts has already run
if ( did_action( 'wp_enqueue_scripts' ) ) {
   $this->enqueue_scripts();
} else {
   add_action( 'wp_enqueue_scripts', [ $this, 'enqueue_scripts' ] );
}
```
Google reCAPTCHA v2 is functioning correctly, I haven't tested v3. You should apply this fix directly to the plugin until an official update is released.

**Comment by MarsWarrior** [comment]
That is great!

So now you see the recaptcha badge?

Most sites at our end use v3. Will check if this fix works. That would make a lot of ppl happy.

**Comment by Gnodesign** [comment]
<img width="400" alt="Image" src="https://github.com/user-attachments/assets/7fee1661-6dbe-4b61-9153-2c810adcb54a" />

@MarsWarrior I tested it using a test key, and v2 works. However, v3 requires a valid key, and I didn't have time to generate one to fully test it. If WPJM's v3 functionality works as expected, this snippet should resolve the issue since the problem is due to the recaptcha script not being loaded.

Let me know once you've tested v3. I've been receiving a lot of reports too, so I had to address this issue.

**Comment by kochhase** [comment]
Hi @MarsWarrior Marswarrior have you tested v3 with this fix from Gnodesign? I will now if other have no problems before i will start with opening the theme.

**Comment by samrueby** [comment]
@Gnodesign's fix works perfectly for v2. Thank you.

**Comment by kochhase** [comment]
> <img alt="Image" width="400" src="https://private-user-images.githubusercontent.com/10162224/420094876-7fee1661-6dbe-4b61-9153-2c810adcb54a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDU1ODg2NjQsIm5iZiI6MTc0NTU4ODM2NCwicGF0aCI6Ii8xMDE2MjIyNC80MjAwOTQ4NzYtN2ZlZTE2NjEtNmRiZS00YjYxLTkxNTMtMmM4MTBhZGNiNTRhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDI1VDEzMzkyNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJlMTM2NzNiZmQyNTkxMWVjMWZkZTk5NjIxMWI5NTFiM2Q2ZWQyNmI2MGU5NDQ0NmM2NzQzNzdiNWE5NGVhZmEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.njH3zGx4km6e_9CtJ0O5k2qMyMAr3gTwTMGXY5txtrA">
> 
> [@MarsWarrior](https://github.com/MarsWarrior) I tested it using a test key, and v2 works. However, v3 requires a valid key, and I didn't have time to generate one to fully test it. If WPJM's v3 functionality works as expected, this snippet should resolve the issue since the problem is due to the recaptcha script not being loaded.
> 
> Let me know once you've tested v3. I've been receiving a lot of reports too, so I had to address this issue.

Hello @Gnodesign Gnodesign Gino, is the fix of the recaptcha problem insite from version 1.9.0? Insite the Update Logs i can not find an mention about the fix. And when will we see support for hCaptcha ??

**Comment by Gnodesign** [comment]
@kochhase No, it is not included in the theme. The fix should be implemented within the plugin itself, but they seem to have gone radio silent. I could submit a PR with the fix if they’re not planning to implement it themselves, but it would most likely be ignored. I’ve submitted several other PRs with fixes and enhancements over the past few months, and unfortunately, there has been no response. 😅

As for hCaptcha, it will most likely be added in version 1.9.2.

**Comment by rndthoughts** [comment]
Posting this for others, in case it helps as we had a similar problem with WP Job Manager "Applications" Plugin.  

We disabled everything on the site apart from WP Jobs Manager and Applications plugins and set it to the WP default theme.  This caused the reCaptcha v2 to render for job application forms.  However, the server let the form submit regardless of whether we checked the "I'm not a robot" box or not.

After a bit of debugging, we ended up modifying this file:

`
/wp-content/plugins/wp-job-manager-applications/includes/class-wp-job-manager-applications-apply.php
`

On line 410 we added:
```
// If reCAPTCHA is enabled and validation fails, throw an exception.
if ( is_wp_error( $recaptcha_result ) ) {
	throw new Exception( $recaptcha_result->get_error_message() );
}
```

This then made the reCaptcha operate correctly in both v2 and v3 modes on the server end

When we re-enabled Divi, we realised it was stripping out the injected Google api.js script, so to get it to appear we added this to the header using Divi -> Integration -> "Add code to the head" in WP Admin:

`<script src="https://www.google.com/recaptcha/api.js" async defer></script>`

We then found v3 failed to work, throwing an error about the wrong site key in the browser console - even though we had the right v3 keys and v3 setting enabled in the Job Manager settings, so we stuck with v2 which worked.

**Comment by samrueby** [comment]
Thank you @rndthoughts

**Comment by Gnodesign** [comment]
@rndthoughts, you can fix it by changing the action on line 25 from:
`add_action( 'loop_start', [ $this, 'form_setup' ] );`
to:
`add_action( 'wp', [ $this, 'form_setup' ] );`

That should make it work without needing any extra code.
